src/sentry/incidents/logic.py

```diff
@@ -105,6 +111,67 @@ def create_incident(
     return incident
 
 
+def calculate_incident_start(query, projects, groups):
+    params = {}
+    if groups:
+        params['issue.id'] = [g.id for g in groups]
+        end = max(g.last_seen for g in groups)
+    else:
+        end = timezone.now()
+
+    params['start'] = end - timedelta(days=7)
+    params['end'] = end
+
+    if projects:
+        params['project_id'] = [p.id for p in projects]
+
+    query_args = get_snuba_query_args(query, params)
+
+    result = raw_query(
+        aggregations=[
+            ('count()', '', 'count'),
+            ('min', 'timestamp', 'first_seen'),
+        ],
+        orderby='time',
+        groupby=['time'],
+        rollup=int(timedelta(minutes=30).total_seconds()),
+        referrer='incidents.calculate_incident_start',
+        limit=10000,
+        **query_args
+    )['data']
+    cur_max = -1
+    cur_start = None
+    max_diff = 0
+    incident_start = None
+    cur_diff = 0
+    prev_count = 0
+    for row in reversed(result):
+        cur_count = row['count']
+        if cur_count < prev_count and cur_count <= cur_max:
+            cur_diff = max(cur_max - cur_count, cur_diff)
+        else:
+            if cur_diff > max_diff:
+                if cur_start:
+                    incident_start = cur_start
```

Comment:
Since this query goes back in time one week, could you ever have two incidents in the last week related with the same group_id ?  let's say you had a big incident 5 days ago, closed 4 days ago and a smaller one yesterday (same issue). It seems this will always return the start time of the first incident.

Reply:
I've done some linear decay of scores just to promote a bit of recency. We'll see how people like this automatic detection.